### ISSUE TITLE
Update and simplify GraphDB repository creation using LoadRDF

### ISSUE THREAD

**Issue by benjamingeer** [dhlab-basel/Knora#550]
Our current Bash scripts for creating GraphDB repositories are based on older versions of GraphDB and the RDF4J console. Let's update this process to use GraphDB's [LoadRDF](http://graphdb.ontotext.com/documentation/standard/loading-data-using-the-loadrdf-tool.html) tool, try to make it simpler for users, and document it better.

**Comment by subotic** [comment]
Would be nice, if it would take a `username` and `password` so that it also works when GraphDB's security is enabled.

**Comment by subotic** [comment]
Ah, LoadRDF is for offline loading.

**Comment by benjamingeer** [comment]
Yes, I hadn’t realised LoadRDF was only for offline loading when I opened this issue. So maybe there’s not much point in doing this.

**Comment by subotic** [comment]
Hmm, maybe we should build the initial setup into `webapi`? Something like, if no repository is found, then create the defined repository with all the basic ontologies and the root user.

After that, it should be easy to load additional data with simple curl requests. Also, I have discovered [Trig](https://www.w3.org/TR/trig/). It's turtle with additional graph information. Basically, the content of the whole triplestore can be downloaded into one file, and also used to load it again into GraphDB, while maintaining consistency.

**Comment by benjamingeer** [comment]
I think as a user, I would rather create the repository myself than have webapi create it automatically for me. There are a lot of GraphDB configuration settings that you can set when you create the repository, and this is something that users may want to customise.

About Trig, yes it looks great, I'm just not sure how widely supported it is by other triplestores.

**Comment by subotic** [comment]
I thought we are talking about the 'make-it-as-simple-as-possible' case. If someone knows what he is doing, then he can go and create the repository by hand (GUI or command line) and tweak whatever he likes. The automatic creation could be a parameter, so only if started with this parameter, `webapi` would create the repository.

Anyway, I want and need something like this :-) Currently, the initial setup requires me to edit the following files:

 - graphdb-knora-prod-data.expect
 - admin-data.ttl
 - permissions-data.ttl

Just to be able to get an empty Knora installation. It also requires me to clone the repository, so that I can actually execute the init script (which I didn't manage to get working with security enabled). I think that this should be way simpler, i.e. start the Knora stack (by using docker-compose) with a custom `webapi` config and then look for the root password in the log output.

Basically, I'm the first real user :-)

About TriG, it shouldn't be difficult to parse it and create a bunch of plain turtle files for each graph. Anyway, I just thought I'll mention it. I will use it for backup, as I need the guaranty to have a consistent copy. Theoretically, it can happen that someone changes something in the time it takes me to download each graph individually, which would require me to somehow lock access.

**Comment by benjamingeer** [comment]
OK, that all sounds good to me!

**Comment by gfoo** [comment]
I've played a little bit with the different containers to prepare a compose for myself (and for other developers here in Lausanne) a kind of ready to use Knora stack for local dev and e2e testing and at this point I failed miserably :-| I don't have enough time now to go deeper and understand why graphdb crashed, why Knora have to wait for Graphdb is running, why it cannot connect to triplestore and so on... 

But a reasonable frontend test env could be : 
- git clone <Knora-test-stack>
- cd  <Knora-test-stack>
- docker-compose up : ready to use grahpdb / sipi / Knora, default ontologies are loaded and so on)
- LoadOntology.sh <folder> : to load my project admin/perm/onto model
- LoadData.sh <folder> : load turtles (dump from graphdb)/ json for POST API/ images 

Anyway, let me know if you need some helps to test/debug your docker-compose mentioned above

**Comment by benjamingeer** [comment]
@gfoo the first thing the Knora API server does when it starts is connect to the triplestore, because without a triplestore it can’t do anything. So the triplestore haa to be started first.

**Comment by benjamingeer** [comment]
@gfoo also it loads all ontologies from the triplestore on startup.

**Comment by gfoo** [comment]
You will face some problems in a docker-compose distributed world, you cannot make any assumptions on the state (not started/crashed/network lags) of the other containers. Knora should run in any cases and at worst the API should return an HTTP 500 error "unreachable triplestore" and retry indefinitely to connect to the triplestore.

**Comment by benjamingeer** [comment]
@gfoo OK, I think we can do that.

**Comment by gfoo** [comment]
prefer HTTP 503, most meaningful in your case : https://fr.wikipedia.org/wiki/Liste_des_codes_HTTP

**Comment by gfoo** [comment]
fix typo in my above message "reasonable frontend" (sorry didn't see my mistake, more understandable I presume now)

**Comment by loicjaouen** [comment]
thanks for answering my late coming questions @benjamingeer, no more questions, but then, anyone: feel free to use my example above if it helps :)

**Comment by subotic** [comment]
I was thinking of adding a route for initializing and loading of data. We shouldn’t allow changing data directly through GraphDB, at least not on our servers.

This will make the startup more simpler. Start `webapi` once and don’t worry. We cannot restart `webapi` in production, just because we loaded some data.

Of course, anybody can do what they want on their own installation.

**Comment by subotic** [comment]
Bulk import is XML, this would be TTL. Basically passing TTL to GraphDB and initiating ontology refresh. We have it already for loading of test data during running of tests, but with authentication and maybe some additional checks.
